File: app/src/main/java/com/salino/sali/data/remote/source/ActivityRemoteDataSource.kt
package com.salino.sali.data.remote.source

import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.firestore.Query
import com.salino.sali.data.model.ActivityLog
import kotlinx.coroutines.tasks.await
import javax.inject.Inject

class ActivityRemoteDataSource @Inject constructor(
    private val firestore: FirebaseFirestore
) {
    private fun activityCollection(householdId: String) = firestore.collection("households")
        .document(householdId)
        .collection("activity")

    fun listenToActivity(householdId: String, onActivity: (List<ActivityLog>) -> Unit): ListenerRegistration {
        return activityCollection(householdId)
            .orderBy("createdAt", Query.Direction.DESCENDING)
            .addSnapshotListener { snapshot, _ ->
                val logs = snapshot?.documents?.mapNotNull { doc ->
                    doc.toObject(ActivityLog::class.java)?.copy(id = doc.id)
                } ?: emptyList()
                onActivity(logs)
            }
    }

    suspend fun upsertActivity(householdId: String, activityLog: ActivityLog) {
        activityCollection(householdId).document(activityLog.id).set(activityLog).await()
    }
}

File: ios/Remote/ActivityRemoteDataSource.swift
import Foundation
import FirebaseFirestore

final class ActivityRemoteDataSource {
    private let firestore: Firestore?

    init(firestore: Firestore? = configuredFirestore()) {
        self.firestore = firestore
    }

    private func activityCollection(_ householdId: String) throws -> CollectionReference {
        guard let firestore else { throw FirebaseConfigurationError.missingGoogleServiceInfo }
        return firestore.collection(FirestoreCollections.households)
            .document(householdId)
            .collection(FirestoreCollections.activity)
    }

    func listenToActivity(householdId: String, onActivity: @escaping ([ActivityLog]) -> Void) -> ListenerRegistration {
        guard let collection = try? activityCollection(householdId) else {
            onActivity([])
            return NoopListenerRegistration()
        }
        return collection
            .order(by: "createdAt", descending: true)
            .addSnapshotListener { snapshot, _ in
                let logs = snapshot?.documents.map { ActivityLog(documentId: $0.documentID, data: $0.data()) } ?? []
                onActivity(logs)
            }
    }

    func upsertActivity(householdId: String, activityLog: ActivityLog) async throws {
        try await activityCollection(householdId).document(activityLog.id).setDataAsync(activityLog.firestoreData)
    }
}
